db/util_testing.go

```diff
@@ -0,0 +1,78 @@
+package db
+
+import (
+	"math"
+
+	"github.com/couchbase/gocb"
+	"github.com/couchbase/sync_gateway/base"
+)
+
+// Workaround SG #3570 by doing a polling loop until the star channel query returns 0 results.
+// Uses the star channel index as a proxy to indicate that _all_ indexes are empty (which might not be true)
+func WaitForIndexEmpty(bucket *base.CouchbaseBucketGoCB, bucketSpec base.BucketSpec) error {
+
+	retryWorker := func() (shouldRetry bool, err error, value interface{}) {
+
+		var results gocb.QueryResults
+
+		// Create the star channel query
+		starChannelQueryStatement := replaceSyncTokensQuery(QueryStarChannel.statement, bucketSpec.UseXattrs)
+		params := map[string]interface{}{}
+		params[QueryParamStartSeq] = 0
+		params[QueryParamEndSeq] = math.MaxInt64
+
+		// Execute the query
+		results, err = bucket.Query(starChannelQueryStatement, params, gocb.RequestPlus, true)
+
+		// If there was an error, then retry.  Assume it's an "index rollback" error which happens as
+		// the index processes the bucket flush operation
+		if err != nil {
+			base.Infof(base.KeyAll, "Error querying star channel: %v.  Assuming it's a temp error, will retry", err)
+			return true, err, nil
+		}
+
+		// Count how many results there are
+		count := CountQueryResults(results)
```

Comment:
Since we don't need an actual count, and just need to know whether it's zero or non-zero, could just call results.Next() once then results.Close.

Reply:
Fixed in 000708f6cd889a112c3fd6e06c49955301ff3297